### PR TITLE
Fix PRS replica state update race condition for collection creation

### PR DESCRIPTION
**Pull request by patsonluk**
*This is a smaller PR copying works from @noblepaul and @chatman in https://github.com/fullstorydev/lucene-solr/pull/168 . Creating this to break the work into smaller pieces which better isolate the problem/solution as well as facilitate easier testing* 😊 

## Description
It is found that when creation PRS collection with 8+ shard, we almost always run into timeout

## Cause
This is triggered by race condition that accidentally set new replica(s) to down states which prevent the PRS collection creation from proceeding:

1. `CreateCollectionCmd` executed by Overseer, it actually submit the create core request to the actual data node via ShardHandler at [here](https://github.com/fullstorydev/lucene-solr/blob/fec2ac23fb5068fddf1a2c3d51956128719ec7a5/solr/core/src/java/org/apache/solr/cloud/api/collections/CreateCollectionCmd.java#L342)
2.  The data node creates the core, first it will publish a `DOWN` state to the core via `ZkController#preRegister` at [here](https://github.com/fullstorydev/lucene-solr/blob/fec2ac23fb5068fddf1a2c3d51956128719ec7a5/solr/core/src/java/org/apache/solr/cloud/ZkController.java#L1905) Take note that this message is submitted to [Overseer Q](https://github.com/fullstorydev/lucene-solr/blob/fec2ac23fb5068fddf1a2c3d51956128719ec7a5/solr/core/src/java/org/apache/solr/cloud/ZkController.java#L1705) since `forcePublish` is `true` . Take note that this send a message to overseer with `operation=state`. Overseer would handle this via `ReplicaMutator#setState` which currently update BOTH the `state.json` and add/update the PRS entries.
3. Next the core creation calls `ZkController#register` which [publish the core state to `ACTIVE`](https://github.com/fullstorydev/lucene-solr/blob/fec2ac23fb5068fddf1a2c3d51956128719ec7a5/solr/core/src/java/org/apache/solr/cloud/ZkController.java#L1378). This triggers a DIRECT PRS entry update at [here](https://github.com/fullstorydev/lucene-solr/blob/fec2ac23fb5068fddf1a2c3d51956128719ec7a5/solr/core/src/java/org/apache/solr/cloud/ZkController.java#L1711)
4. Now Overseer will be waiting at [here](https://github.com/fullstorydev/lucene-solr/blob/fec2ac23fb5068fddf1a2c3d51956128719ec7a5/solr/core/src/java/org/apache/solr/cloud/api/collections/CreateCollectionCmd.java#L353) for all the core PRS state to be active. If the `DOWN` message sent to overseer is processed before step 3, then it's fine. However, when there are more shards, it's likely that there are delays in step 2, such that the `DOWN` on replica from step 2 overwrites the `ACTIVE` in step 3. Hence it will eventually timeout

## Solution
This is caused by race condition:
1. Overseer handling of `operation=state` message which triggers `ReplicaMutator#setState` which updates BOTH the `state.json` and adds/updates the PRS entries - `DOWN` PRS entry
2. `ZkController#register` (from the data node that creates the new core) tries to update the PRS entries directly - `ACTIVE` PRS entry

Take note that point 1. violates the higher level design that: **PRS entry should be updated by data node only**, which such design should avoid this type of race conditional

Therefore @noblepaul and @chatman applied fixes of:
1. Removed the PRS specific mutation in `ReplicaMutator#setState` (which i slightly simplified to remove the extra flag)
2. Ensure `ZkController#publish` is the one that handles PRS entries for PRS enabled collections. Hence it will always write the PRS entry.

Take note that this modifies the expected behavior of overseer's handling on `operation:state` message. Before it is expected to BOTH update the `state.json` and the `PRS entries` and now it ONLY modify the `state.json`. Whoever submit such message to overseer needs to be aware of the extra work for PRS. From references searches, `ZkController#publish` appears to be the only trigger point

## Tests
Tested on playpen:
1. On PRS collection `o-HNS-na1`, split and move operation to make sure it's working
2. Create a PRS collection with 1024 shards: 
`curl "http://localhost:8985/solr/admin/collections?action=CREATE&name=test-prs-creation&numShards=1024&replicationFactor=1&wt=json&collection.configName=_FS4&router.name=compositeId&maxShardsPerNode=1024&perReplicaState=true"` . The creation was successful. 
```
(CONNECTED [localhost:2181]) /solr/collections/test-prs-creation> stat state.json
Stat(
  czxid=0xa9000458a9
  mzxid=0xa90004af41
  ctime=1659656239374
  mtime=1659656328236
  version=2048
  cversion=7168
  aversion=0
  ephemeralOwner=0x0
  dataLength=305685
  numChildren=1024
  pzxid=0xa90004b0ae
)
```
shows the state.json was updated 2048 times - expected for each replica: 
- +1 from CreateCollectionCmd for each replicaPosition
- +1 from DOWN state publish. 
This will be improved in PR #173 

As for child updates 7168 times - expected for each replica,
- +1 for ZkController#preRegister, in ZkController#publish, direct write PRS to down
- +2 for Overseer handling on "operation=state", ZkWriteCommand calls PerReplicaStatesOps#touchChildren
- +2 for runLeaderProcess, flip the replica to leader
- +2 for ZkController#register, in ZkController#publish, direct write PRS to active
Take note that +1 for ADD, and +2 for all the UPDATE (remove the old PRS and add new PRS entry)

## TODO
Update the unit test cases, some of them might start failing as the expected behavior of overseer's handling on `operation:state` message has now changed. **update: no new test case failures**

### Other race conditions
Take note that there are other race conditions with very similar pattern (ie overseer process message via Q `ClusterStateUpdater` vs  direct write either from overseer or data node. Another example described in [setting shard leader race condition](https://docs.google.com/document/d/17fCMTMV3P1XoB8HatbVsuwavuOorAVLdOinCR6k_HYs/edit). Such access usually are reflected in `SliceMutator` and `ReplicaMutator`. From my observation, the design of those 2 Mutators are originally for Overseer only, which figures out the `ZkWriteCommand` and then execute on the overseer node. However since PRS, those mutators seem to be used to "figure out the updated doc collection state" which the ZkCommand is actually not used for any zk operations, for example [this](https://github.com/fullstorydev/lucene-solr/blob/fec2ac23fb5068fddf1a2c3d51956128719ec7a5/solr/core/src/java/org/apache/solr/cloud/api/collections/CreateCollectionCmd.java#L280). 

However this makes it a bit tricky to fix race condition. I would advise (IN THE FUTURE!!! not in this PR):
1. Make `SliceMutator` and `ReplicaMutator` purely something used by overseer for Zk operations, by doing this we can remove all the PRS entries operation inside. The benefit of this is we can ensure overseer NEVER touch the PRS entries
3. Look for references that call those 2 mutators, see if those callers need to handle the PRS entries themselves. Much like in this PR, `ZkController#publish` would send the `operation=state` message to Overseer (which eventually trigger `RepliacMutator#setState`, such publish logic needs to do its own PRS operations instead o relying of the Mutator (which should be ONLY used by overseer)

### Not yet fixing ALL prs entry race condition
Bad naming on the branch, this fixes race condition between `zkController#publish` and `ReplicaMutator`, which is sorta related to add replica. But add replica also has other ops that updates the PRS entries directly in `SliceMutator`, such is not addressed in this PR.

Both `CreateCollectionCmd` and `AddReplicaCmd` eventually call `SliceMutator#addReplica` which still does PRS entries update by embedding it in the `ZkWriteCommand `it returns. It is okay for collection creation cmd the `ZkWriteCommand` returned was basically ignored :joy_cat: (not getting executed) except that command.collection is extracted. For `AddReplicaCmd`  tho, the `ZkWriteComand` is executed in OS, which means OS is still writing PRS entries

### CODE REVIEW

File: solr/core/src/java/org/apache/solr/cloud/ZkController.java [hunk 1]
```diff
@@ -1702,8 +1702,13 @@ public void publish(final CoreDescriptor cd, final Replica.State state, boolean
...
       DocCollection coll = zkStateReader.getCollection(collection);
       if (forcePublish || sendToOverseer(coll, coreNodeName)) {
+        //for PRS, this only updates state.json but not the actual PRS entries
         overseerJobQueue.offer(Utils.toJSON(m));
-      } else {
+      }
+
+      //extra handling for PRS, we need to write the PRS entries from this node directly,
+      //as overseer does not and should not handle those entries
+      if (coll != null && coll.isPerReplicaState() && coreNodeName != null) {
```
Comment (hiteshk25) [inline solr/core/src/java/org/apache/solr/cloud/ZkController.java:1711]: @patsonluk here we are might update state.json through overseer above. Likely those will be shard's state or other params. But we need to make sure race condition between these two operations(overseer state.json update vs direct prs) should not cause any other issue.

Comment (patsonluk) [reply solr/core/src/java/org/apache/solr/cloud/ZkController.java:1711]: Race conditions that concern me so far are the state.json vs state.json or PRS entries vs PRS entries. Do you think the ordering of update on state.json vs PRS entries would cause any issues? :)

Comment (hiteshk25) [reply solr/core/src/java/org/apache/solr/cloud/ZkController.java:1711]: The state.json maintains collection structure and prs state maintains replica-status. So better to keep this semantics. 

Also, good to see if we don't change this then what fails!

Comment (patsonluk) [reply solr/core/src/java/org/apache/solr/cloud/ZkController.java:1711]: Could probably tell by inspecting the loading logic of replica states on PRS (will do it tomorrow 😉 ) but my guess is that it should be okay - it will either temporarily not have the replica in cluster state yet or have a replica but the replica state (leadership/active/inactive) temporarily wrong. If the whatever logic that reads such state does expect and handle temporary inconsistent replica state gracefully (for example just wait and check again as in `CreateCollectionCmd`, then it will likely be okay eventually

Comment (noblepaul) [reply solr/core/src/java/org/apache/solr/cloud/ZkController.java:1711]: agree with @patsonluk here. This is unlikely to cause any inconsistency.

File: solr/core/src/java/org/apache/solr/cloud/overseer/ReplicaMutator.java [hunk 2]
```diff
@@ -252,7 +252,6 @@ private ZkWriteCommand updateState(final ClusterState prevState, ZkNodeProps mes
       log.info("Failed to update state because the replica does not exist, {}", message);
       return ZkStateWriter.NO_OP;
     }
-    boolean persistCollectionState = collection != null && collection.isPerReplicaState();
```
Comment (noblepaul) [inline solr/core/src/java/org/apache/solr/cloud/overseer/ReplicaMutator.java:255]: Why is this removed?

If we remove this, it will result in unnecessary updates to `state.json`

Comment (patsonluk) [reply solr/core/src/java/org/apache/solr/cloud/overseer/ReplicaMutator.java:255]: @noblepaul from my understanding of the current release/8.8 [`ReplicaMutator`](https://github.com/fullstorydev/lucene-solr/blob/fec2ac23fb5068fddf1a2c3d51956128719ec7a5/solr/core/src/java/org/apache/solr/cloud/overseer/ReplicaMutator.java#L255), it appears that:
1. For NON prs operation, this flag does not matter, it always return the ZkWriteCmd
2. For PRS operation, this flag is always true unless "collection is null" (which means someone is submitting `operation=state` message to OS Q when the collection is not yet exists - think it's not a valid case?)

Since in this current PR, we are correcting the logic of `ReplicaMutator#updateState` to ONLY perform `state.json` operation (no more PRS ops), both PRS and non PRS operation should return the same `ZkWriteCommand`, and such `persistCollectionState` flag does not have any reference anymore.

Please lemme know if any cases are overlooked 😄  . My assumption (which I will add javadoc) to this method now, is that it should only update state.json. If any initiator of such update (usually whoever submits the `operation=state` to OS Q) should be aware of exactly what it's trying to update. If it's PRS collection, it should determine if such update would affect PRS entries, state.json or both and do the corresponding actions accordingly (send `operation=state` to OS Q and/or update the PRS entries directly).

Comment (noblepaul) [reply solr/core/src/java/org/apache/solr/cloud/overseer/ReplicaMutator.java:255]: > Since in this current PR, we are correcting the logic of ReplicaMutator#updateState to ONLY perform state.json operation (no more PRS ops), both PRS and non PRS operation should return the same ZkWriteCommand

That is not true. The state message can change the `state` flag which results in modification of `state.json` and that is unnecessary for PRS collections.

Comment (hiteshk25) [reply solr/core/src/java/org/apache/solr/cloud/overseer/ReplicaMutator.java:255]: @noblepaul @patsonluk  would be good to see unit test for it first.

Comment (noblepaul) [reply solr/core/src/java/org/apache/solr/cloud/overseer/ReplicaMutator.java:255]: @hiteshk25 there is a [testcase](https://github.com/fullstorydev/lucene-solr/blob/3a8818ca1fb209156417ed18ccc3cdabc1edf260/solr/core/src/test/org/apache/solr/core/FSPRSTest.java#L333) already in this #168 . The test ensures that `state.json` is not modified unenecessarily

Comment (patsonluk) [reply solr/core/src/java/org/apache/solr/cloud/overseer/ReplicaMutator.java:255]: > > Since in this current PR, we are correcting the logic of ReplicaMutator#updateState to ONLY perform state.json operation (no more PRS ops), both PRS and non PRS operation should return the same ZkWriteCommand
> 
> That is not true. The state message can change the `state` flag which results in modification of `state.json` and that is unnecessary for PRS collections.

In the existing code as far as `collection.isPerReplicaState()` is `true`, then such flag can only be `true` (since all the other assignment assigns `true`), that means the existing code seems to always updates `state.json` for PRS collection, is that a behavior we want to change as a part of this fix? 

Are there any special cases for PRS that we should treat this call differently from the existing non PRS flow? I kinda like the `ReplicateMutator` code being generic and not have to be "PRS aware" 😊  . On the other end whoever is going to make replica state changes (for example `ZkController#publish` as in your change 😄 ) should and need to have the knowledge of "doing the right thing" - ie it should only send `State` message to overseer if it wants state.json update, and it should do PRS ops separately if it requires PRS entries update directly. This will likely become a new requirement if we want PRS entries update to be only taken care of by the data node code. (ie no update from overseer)

Comment (noblepaul) [reply solr/core/src/java/org/apache/solr/cloud/overseer/ReplicaMutator.java:255]: >hat means the existing code seems to always updates `state.json` for PRS collection, is that a behavior we want to change as a part of this fix?

Yes. absolutely. In a PRS collection, there is no point in updating the `state` flag of a replica in `state.json` as it is always overridden by the value in the individual replica states

Comment (patsonluk) [reply solr/core/src/java/org/apache/solr/cloud/overseer/ReplicaMutator.java:255]: Thanks for the explanation @noblepaul ! 💡 My original intention of this PR is extract from #168 to focus on the race condition - enforcing PRS ops in data node and remove it from overseer replica mutator, and leave the rest of the behaviors exactly the same as before. 

I'm okay if we want to change this as a part of the PR too, that appears to be an optimization (fixing redundant state.json updates as in the existing code?). what do u think @hiteshk25 😊

Comment (patsonluk) [reply solr/core/src/java/org/apache/solr/cloud/overseer/ReplicaMutator.java:255]: Went into a bit more details this morning to assess the scope of the redundant state.json writes for our common usages in FS. I have found:
- Collection Creation
- Add replica (solrman move/split operation)

Since both of the above creates a new core which triggers `ZkController#publish` [with `forcePublish` as `true`](https://github.com/fullstorydev/lucene-solr/blob/fec2ac23fb5068fddf1a2c3d51956128719ec7a5/solr/core/src/java/org/apache/solr/cloud/ZkController.java#L1905), which indeed would issue a `state.json` update when replica state is set as `DOWN` . Such update is indeed redundant/unnecessary as @noblepaul pointed out.

Out of the above cases, the redundant update on `state.json` for collection creation could be more of a concern (as batching is not enabled in ZkStateWriter PRS collection), therefore creating a PRS collection with n shard could potentially issue n unnecessary writes to state.json?

The redundant state.json write however, does not affect:
- Solr service stop (in our case SIGTERM, if unsuccessful SIGKILL)
- Solr service start (`ZkController#preRegister` this [condition](https://github.com/fullstorydev/lucene-solr/blob/fec2ac23fb5068fddf1a2c3d51956128719ec7a5/solr/core/src/java/org/apache/solr/cloud/ZkController.java#L1904) is `false`, then on `ZkController#register` state `ACTIVE`, `forcePubish` and `sendToOverseer` are `false` at [here](https://github.com/fullstorydev/lucene-solr/blob/fec2ac23fb5068fddf1a2c3d51956128719ec7a5/solr/core/src/java/org/apache/solr/cloud/ZkController.java#L1704) hence no message to OS)

Comment (noblepaul) [reply solr/core/src/java/org/apache/solr/cloud/overseer/ReplicaMutator.java:255]: > herefore creating a PRS collection with n shard could potentially issue n unnecessary writes to state.json? 

true

> Solr service start (ZkController#preRegister this [condition](https://github.com/fullstorydev/lucene-solr/blob/fec2ac23fb5068fddf1a2c3d51956128719ec7a5/solr/core/src/java/org/apache/solr/cloud/ZkController.java#L1904) is false, then on ZkController#register state ACTIVE, forcePubish and sendToOverseer are false at [here](https://github.com/fullstorydev/lucene-solr/blob/fec2ac23fb5068fddf1a2c3d51956128719ec7a5/solr/core/src/java/org/apache/solr/cloud/ZkController.java#L1704) hence no message to OS)

True. But for our JUnit tests, restart always start the node in another port and ends up changing the `node_name` & `base_url` attributes

**Comment by noblepaul** [comment]
@patsonluk 

Can you please copy [this test](https://github.com/fullstorydev/lucene-solr/blob/bd1716d4b5d42261d7f701667f27b494896bc845/solr/core/src/test/org/apache/solr/core/FSPRSTest.java#L333) to the PR

File: solr/core/src/java/org/apache/solr/cloud/ZkController.java (hunk 1 above)
Comment (patsonluk) [reply solr/core/src/java/org/apache/solr/cloud/ZkController.java:1711]: @hiteshk25 As discussed, your concern is that this is a "change of behavior", as before this change, the `state.json` update AND the PRS entries update (if `forcePublish || sendToOverseer(coll, coreNodeName)` evaluates to `true`) happen at the same piece of code (`ReplicaMutator#updateState`) and in predictable execution order - state.json update first then PRS entries. 

With the code change here, it's very likely that PRS entries update would happen BEFORE the state.json update, and the ordering is not deterministic.

Hence the only weaker guarantee we have with this new change, is that "eventually" the PRS and state.json will both be updated.

For the several places that watch for replica state changes, they seem to be okay with this weaker guarantee, such as [here](https://github.com/fullstorydev/lucene-solr/blob/fec2ac23fb5068fddf1a2c3d51956128719ec7a5/solr/solrj/src/java/org/apache/solr/common/cloud/ZkStateReader.java#L1881) in Solr java code and [here](https://github.com/cowpaths/mn/blob/master/projects/fullstory/go/src/fs/services/solrman/internal/solrmansvc/move.go#L112) in our go code

Can you check if any code u know would have issue with such weaker guarantee?

I agree that ideally we want to keep the deterministic execution order as before, however it could be a bit challenging to "block" and wait for the completion of state.json update in OS before proceeding, as  `waitForState` probably wouldn't work for PRS collection replica entry update as it will not be able to see the change in PRS state after sending it to OS as we now move PRS entry operations away from the overseer handling.  😓

**Comment by patsonluk** [comment]
> @patsonluk
> 
> Can you please copy [this test](https://github.com/fullstorydev/lucene-solr/blob/bd1716d4b5d42261d7f701667f27b494896bc845/solr/core/src/test/org/apache/solr/core/FSPRSTest.java#L333) to the PR

Great test there! I will add the `testZkNodeVersions`. By inspection I think the test case will not match the `assertEquals(1, stat.getVersion());` for PRS operation. (Changed the expect value in the test case, added breakdown on the count as comment in the test case)

I agree that ideally we do want to eventually drive the state.json update to once only for PRS collection (and probably less for non PRS collection too). However, for this PR, we really want to focus on fixing the PRS entry state race condition. 

Therefore I propose keeping the change simple as it is now. 

My thoughts:
1. We can consider the optimization to NOT update state.json on PRS entries update later on - this current change does not make anything worse. The # of state.json update is same as before the change
2. As for the optimization itself, I think we need more time to go through the options. I believe in the #168, it adds some extra logic in the `ReplicaMutator` to check whether state.json needs to be update, while it likely will work fine for our case, I personally find that pretty hard to follow - even as someone that now has some experience with PRS feature. This might be very confusing to other dev that has less exposure to PRS. I don't want to eliminate the option to modify on the caller side instead `preRegister`/`ZkController#publish` which it might have a better idea to tell whether an update the state.json is truly necessary, such that it wouldn't push a `state` message to the OS queue and ended up being a no-op. Honestly, I don't know which way is better/easier (or maybe there are some even better ways). But I think we need more time for that and it shouldn't be a blocker of this fix as is.
3. We have already cut the state.json update by half with the other change you proposed, which we extracted to #173 . So the bottom line - we should already be in a better position than before

File: solr/core/src/java/org/apache/solr/cloud/ZkController.java (hunk 1 above)
Comment (patsonluk) [reply solr/core/src/java/org/apache/solr/cloud/ZkController.java:1711]: As discussed with @hiteshk25, going to document the findings here:

Our concern for this code is that if state.json update and PRS update are not in the correct order (the risk described in next comment, too long so I separated it out). Then we could risk the PRS update being dropped by our go solr code.

However, if `state.json` already has the replica in the json content, then it's relatively safer. Hence, we are going to assess operations that we care, and see if `state.json` already knows about the replica:

1. Split/move shard - which eventually call ADDREPLICA. It is safe as it first call `SliceMutator#addReplica` which [translate to a ZkWriteCommand that updates the `state.json`](https://github.com/fullstorydev/lucene-solr/blob/42259c6af2a1f57ff008663395e3b90936b1a3e3/solr/core/src/java/org/apache/solr/cloud/overseer/SliceMutator.java#L101). Then it will [wait](https://github.com/fullstorydev/lucene-solr/blob/42259c6af2a1f57ff008663395e3b90936b1a3e3/solr/core/src/java/org/apache/solr/cloud/api/collections/AddReplicaCmd.java#L265) to see the replica in place first before proceeding. Later on the `ZkController#preRegister` will have `forcePublish` as `true`
2. Create collection - Also safe as `CreateCollectionCmd` [updates the `state.json`](https://github.com/fullstorydev/lucene-solr/blob/42259c6af2a1f57ff008663395e3b90936b1a3e3/solr/core/src/java/org/apache/solr/cloud/api/collections/CreateCollectionCmd.java#L282) before the cores are created (which later call `ZkController#preRegister` which triggers this code

Comment (patsonluk) [reply solr/core/src/java/org/apache/solr/cloud/ZkController.java:1711]: **This comment is below solely for documentation purpose. NOT that important for current situation**

it could be a bit problematic if PRS entry is updated BEFORE the corresponding replica is added to `state.json`. 

Normally, when the replica details is added to state.json first. it would trigger the "data notification" with the state.json as bytes, such byte array will be parsed and fill up the `collection` struct (which has collection -> shard -> replica info). For PRS collection, the replica info will then be ["replace" by](https://github.com/fullstorydev/gosolr/blob/965388d4e5e635f8984962fa4a611a642b0da283/solrmonitor/solrmonitor.go#L622) whatever is in the old/existing `collection` struct, last observed (if existed) replica state triggered by child watch. Then when the PRS entries are updated, it will trigger the "child watch" with the PRS entry string, it will then [update](https://github.com/fullstorydev/gosolr/blob/965388d4e5e635f8984962fa4a611a642b0da283/solrmonitor/solrmonitor.go#L278) the replica info in the `collection` struct.

Now consider the cases that IF PRS entry is updated BEFORE the corresponding replica is added: The child watch routine will now not do anything as in the [code](https://github.com/fullstorydev/gosolr/blob/965388d4e5e635f8984962fa4a611a642b0da283/solrmonitor/solrmonitor.go#L278) linked earlier, if such replica does not existing in the `collection` struct yet, the update will be skipped.  Then when the state.json is updated, IF somehow the state.json has matching replica details as the PRS entry, then *it will be okay*. It will find no previous replica details to replace with and just use the details in  the incoming state.json. However, if the state.json does not contain the correct replica details (active/leadership, and possible more details in the future), then the replica details in `collection` struct will be wrong.

This is probably not an issue in `ZkController#publish` for now, as the 2 operations of concern:
1. state.json update : `overseerJobQueue.offer(Utils.toJSON(m));`, which `m` is derived from `props.put(ZkStateReader.STATE_PROP, state.toString());`
2. PRS entry update: `PerReplicaStatesOps.flipState(coreNodeName, state, perReplicaStates)`

So they uses the same `state` object hence should be using the same data.

It could be problematic if those 2 pieces of details diverge in the future (for example certain replica details are only kept in PRS entry and not in state.json anymore) or any bugs that causes any discrepancies.

**Review by noblepaul — APPROVED** [review]
LGTM

**Review by hiteshk25 — APPROVED** [review]
LGTM
